src/sage/rings/function_field/drinfeld_modules/charzero_drinfeld_module.py

```diff
@@ -0,0 +1,418 @@
+            sage: log = phi.logarithm(); log
+            z + ((1/(T^2+T))*z^2) + ((1/(T^6+T^5+T^3+T^2))*z^4) + O(z^8)
+
+        The logarithm is returned as a lazy power series, meaning that
+        any of its coefficients can be computed on demands::
+
+            sage: log[2^4]
+            1/(T^30 + T^29 + T^27 + ... + T^7 + T^5 + T^4)
+            sage: log[2^5]
+            1/(T^62 + T^61 + T^59 + ... + T^8 + T^6 + T^5)
+
+        Example in higher rank::
+
+            sage: A = GF(5)['T']
+            sage: K.<T> = Frac(A)
+            sage: phi = DrinfeldModule(A, [T, T^2, T + T^2 + T^4, 1])
+            sage: phi.logarithm()
+            z + ((4*T/(T^4+4))*z^5) + O(z^8)
+
+        TESTS::
+
+            sage: A = GF(2)['T']
+            sage: K.<T> = Frac(A)
+            sage: phi = DrinfeldModule(A, [T, 1])
+            sage: q = 2
+            sage: log[2] == -1/((T**q - T))  # expected value
+            True
+            sage: log[2**2] == 1/((T**q - T)*(T**(q**2) - T))  # expected value
+            True
+            sage: log[2**3] == -1/((T**q - T)*(T**(q**2) - T)*(T**(q**3) - T))  # expected value
+            True
+
+        """
+        L = LazyPowerSeriesRing(self._base, name)
+        zero = self._base.zero()
+        q = self._Fq.cardinality()
+
+        def coeff_log(k):
+            # Return the k-th coefficient of the logarithm
+            k = ZZ(k)
+            if k.is_power_of(q):
+                return self._compute_coefficient_log(k.log(q))
+            else:
+                return self._base.zero()
+        return L(coeff_log, valuation=1)
+
+    @cached_method
+    def _compute_goss_polynomial(self, n, q, poly_ring, X):
+        r"""
+        Utility function for computing the n-th Goss polynomial.
+
+        The user should not call this method directly, but
+        :meth:`goss_polynomial` instead.
+
+        TESTS::
+
+            sage: A = GF(2^2)['T']
+            sage: K.<T> = Frac(A)
+            sage: phi = DrinfeldModule(A, [T, T+1, T^2, 1])
+            sage: poly_ring = phi.base()['X']
+            sage: X = poly_ring.gen()
+            sage: phi._compute_goss_polynomial(0, 2^2, poly_ring, X)
+            0
+            sage: phi._compute_goss_polynomial(3, 2^2, poly_ring, X)
+            X^3
+            sage: phi._compute_goss_polynomial(4*3, 2^2, poly_ring, X)
+            X^12
+            sage: phi._compute_goss_polynomial(9, 2^2, poly_ring, X)
+            X^9 + (1/(T^3 + T^2 + T))*X^6 + (1/(T^6 + T^4 + T^2))*X^3
+
+        """
+        # Trivial cases
+        if n.is_zero():
+            return poly_ring.zero()
+        if n <= q - 1:
+            return X**n
+        if n % q == 0:
+            return self.goss_polynomial(n // q)**q
+        # General case
+        pol = sum(self._compute_coefficient_exp(i+1)
+                  *self._compute_goss_polynomial(n - q**(i+1), q, poly_ring, X)
+                  for i in range(0, (n.log(q).n()).floor()))
```

Comment:
I think it can cause problems if `n.log(q)` is an integer but `n.log(q).n()` is a little bit smaller than its exact value.

Reply:
I agree, but by to the properties of Goss polynomials, especially this one: $G_{qn} = G_{n}^q$, and by the recursion, this loop will only be evaluated when $n = q$ in which case we have `n.log(q).n() = 1.0000000`. Hence, I'm not sure if this problem will occur.

Reply:
Why do you need a loop then? Can't you just write:
```
pol = self._compute_coefficient_exp(1) * self._compute_goss_polynomial(n - q, q, poly_ring, X)
```
It seems that I'm missing something.

Reply:
By the way, I think it's better (and probably faster) to use the `val_unit` method. (See my next comments.)